cmd/clusterctl/cmd/migrate.go

```diff
@@ -0,0 +1,189 @@
+/*
+Copyright 2025 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/core/v1beta2"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/internal/migrate"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/internal/scheme"
+)
+
+type migrateOptions struct {
+	output    string
+	toVersion string
+}
+
+var migrateOpts = &migrateOptions{}
+
+var supportedTargetVersions = []string{
+	clusterv1.GroupVersion.Version,
+}
+
+var migrateCmd = &cobra.Command{
+	Use:   "migrate [SOURCE]",
+	Short: "EXPERIMENTAL: Migrate cluster.x-k8s.io resources between API versions",
+	Long: `EXPERIMENTAL: Migrate cluster.x-k8s.io resources between API versions.
+
+This command is EXPERIMENTAL and may be removed in a future release!
+
+Scope and limitations:
+- Only cluster.x-k8s.io resources are converted
+- Other CAPI API groups are passed through unchanged
+- ClusterClass patches are not migrated
+- Field order may change and comments will be removed in output
+- API version references are dropped during conversion (except ClusterClass and external
+  remediation references)
+
+Examples:
+  # Migrate from file to stdout
+  clusterctl migrate cluster.yaml
+
+  # Migrate from stdin to stdout
+  cat cluster.yaml | clusterctl migrate
+
+  # Explicitly specify target <VERSION>
+  clusterctl migrate cluster.yaml --to-version <VERSION> --output migrated-cluster.yaml`,
+
+	Args: cobra.MaximumNArgs(1),
+	RunE: func(_ *cobra.Command, args []string) error {
+		return runMigrate(args)
+	},
+}
+
+func init() {
+	migrateCmd.Flags().StringVarP(&migrateOpts.output, "output", "o", "", "Output file path (default: stdout)")
+	migrateCmd.Flags().StringVar(&migrateOpts.toVersion, "to-version", clusterv1.GroupVersion.Version, fmt.Sprintf("Target API version for migration (supported: %s)", strings.Join(supportedTargetVersions, ", ")))
+
+	RootCmd.AddCommand(migrateCmd)
+}
+
+func isSupportedTargetVersion(version string) bool {
+	for _, v := range supportedTargetVersions {
+		if v == version {
+			return true
+		}
+	}
+	return false
+}
+
+func runMigrate(args []string) error {
```

Comment:
Most of the code in this func must be moved to the client interface method, all except:
- flag validation
- input handling (reading from a file or stdin)
- output handling (writing to file, stdout, stderr)

Reply:
Addressed this in the upcoming refactoring.